pgrx/src/spi.rs

```diff
@@ -409,25 +409,26 @@ impl<'a> Query for &'a str {
 }
 
 #[derive(Debug)]
-pub struct SpiTupleTable {
+pub struct SpiTupleTable<'conn> {
     #[allow(dead_code)]
     status_code: SpiOkCodes,
-    table: Option<*mut pg_sys::SPITupleTable>,
+    table: Option<&'conn mut pg_sys::SPITupleTable>,
```

Comment:
Do we later use this to access data allocated past the SPITubleTable, or does it alias with anything else? I notice it doesn't get a lot of action (directly) in our test infra, so I'm wondering if the new `&'conn mut` will break existing user code.

Reply:
Being an implementation detail, I'm unclear how it would break existing user code?

If that thing were to leak out to the public API tho, at least now it has a proper lifetime and might affect compilation, which we'd want, yeah?

Reply:
There's a few details where internal changes can actually leak out due to Rust unfortunately inferring certain information that is otherwise fairly laborious to annotate on _every single type ever_, and I can never off-handedly remember exactly which things get affected by this. I think `*mut T` to `&'a mut T` is usually nonbreaking, however, it's `&'a T` to `&'a mut T` that reliably gotchas people.

This is sometimes referred to as the "perfect derive problem" or "semver hazards": https://github.com/rust-lang/lang-team/blob/master/design-meeting-minutes/2022-04-12-implied-bounds-and-perfect-derive.md

Reply:
Jubilee, is there an action item here?  I’m not really sure.

Reply:
No, I don't think so, I'm just noting it as something to keep an eye on and to try to write more tests that exercise the more niche points of `pgrx::spi`.

Reply:
Well, there was the question about aliasing as usual, and also about range of access, given that I'm not immediately sure if `pg_sys::SPITubleTable` is like `ArrayType` or not.

Reply:
Range of access doesn't really matter. Stacked borrowed forbids it, but treed-borrows allows `&mut Header` with trailing data, and I've been told that even though TB is likely to be replaced, that problem will certainly remain fixed in the replacement.

IOW, this only matters for running in miri (under default settings), which isn't a concern.